enhancements/20241211-upgrade-enhancement.md

```diff
@@ -0,0 +1,997 @@
+	State   string `json:"state,omitempty"`
+	Reason  string `json:"reason,omitempty"`
+	Message string `json:"message,omitempty"`
+}
+
+// UpgradePlanPhase defines what overall phase UpgradePlan is in
+type UpgradePlanPhase string
+
+type UpgradePlanPhaseTransitionTimestamp struct {
+	Phase                    UpgradePlanPhase `json:"phase"`
+	PhaseTransitionTimestamp metav1.Time      `json:"phaseTransitionTimestamp"`
+}
+
+type ReleaseMetadata struct {
+	Harvester            string `json:"harvester,omitempty"`
+	HarvesterChart       string `json:"harvesterChart,omitempty"`
+	OS                   string `json:"os,omitempty"`
+	Kubernetes           string `json:"kubernetes,omitempty"`
+	Rancher              string `json:"rancher,omitempty"`
+	MonitoringChart      string `json:"monitoringChart,omitempty"`
+	MinUpgradableVersion string `json:"minUpgradableVersion,omitempty"`
+}
+
+// UpgradePlanSpec defines the desired state of UpgradePlan
+type UpgradePlanSpec struct {
+	// version refers to the corresponding version resource in the same namespace.
+	// +required
+	Version string `json:"version"`
+
+	// upgrade can be specified to opt for any other specific upgrade image. If not provided, the version resource name is used.
+	// For instance, specifying "dev" for the field can go for the "rancher/harvester-upgrade:dev" image.
+	// +optional
+	Upgrade *string `json:"upgrade,omitempty"`
+
+	// force indicates the UpgradePlan will be forcibly applied, ignoring any pre-upgrade check failures. Default to "false".
+	// +optional
+	Force *bool `json:"force,omitempty"`
+
+	// mode represents the manipulative style of the UpgradePlan. Can be either of "automatic" or "interactive". Default to "automatic".
+	// +optional
+	Mode *string `json:"mode,omitempty"`
+}
+
+// UpgradePlanStatus defines the observed state of UpgradePlan.
+type UpgradePlanStatus struct {
+	// conditions represent the current state of the UpgradePlan resource.
+	// Each condition has a unique type and reflects the status of a specific aspect of the resource.
+	//
+	// The status of each condition is one of True, False, or Unknown.
+	// +listType=map
+	// +listMapKey=type
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+	// isoImageID refers to the namespaced name of the VM image that will be used for the upgrade.
+	// +optional
+	ISOImageID *string `json:"isoImageID,omitempty"`
+
+	// nodeStatuses reflect each node's upgrade status for node specific tasks.
+	// +mapType=atomic
+	// +optional
+	NodeUpgradeStatuses map[string]NodeUpgradeStatus `json:"nodeUpgradeStatuses,omitempty"`
+
+	// phase shows what overall phase the UpgradePlan resource is in.
+	Phase UpgradePlanPhase `json:"phase,omitempty"`
+
+	// phaseTransitionTimestamp is the timestamp of when the last phase change occurred.
+	// +listType=atomic
+	// +optional
+	PhaseTransitionTimestamps []UpgradePlanPhaseTransitionTimestamp `json:"phaseTransitionTimestamps,omitempty"`
```

Comment:
`Phase` is already captured in `PhaseTransitionTimestamps`.

Reply:
Exactly. Or do you suggest renaming it to `CurrentPhase`? I simply followed the naming from the VirtualMachineInstanceMigration CRD.